packages/eslint-plugin/src/rules/unbound-method.ts

```diff
@@ -0,0 +1,132 @@
+/**
+ * @fileoverview Enforces unbound methods are called with their expected scope.
+ * @author Josh Goldberg <https://github.com/joshuakgoldberg>
+ */
+'use strict';
+
+import { TSESTree } from '@typescript-eslint/typescript-estree';
+import * as tsutils from 'tsutils';
+import * as ts from 'typescript';
+
+import * as util from '../util';
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+interface Config {
+  ignoreStatic: boolean;
+}
+
+type Options = [Config];
+
+type MessageIds = 'unbound';
+
+/**
+ * @type {import("eslint").Rule.RuleModule}
+ */
+export default util.createRule<Options, MessageIds>({
+  name: 'unbound-method',
+  meta: {
+    docs: {
+      category: 'Best Practices',
+      description:
+        'Enforces unbound methods are called with their expected scope.',
+      tslintName: 'no-unbound-method',
+      recommended: 'error'
+    },
+    messages: {
+      unbound:
+        'Avoid referencing unbound methods which may cause unintentional scoping of `this`.'
+    },
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          ignoreStatic: {
+            type: 'boolean'
+          }
+        },
+        additionalProperties: false
+      }
+    ],
+    type: 'problem'
+  },
+  defaultOptions: [
+    {
+      ignoreStatic: false
+    }
+  ],
+  create(context, [{ ignoreStatic }]) {
+    const parserServices = util.getParserServices(context);
+    const checker = parserServices.program.getTypeChecker();
+
+    return {
+      MemberExpression(node) {
+        if (isSafeUse(node)) {
+          return;
+        }
+
+        const originalNode = parserServices.esTreeNodeToTSNodeMap.get(node);
+        const symbol = checker.getSymbolAtLocation(originalNode);
+
+        if (symbol && isDangerousMethod(symbol, ignoreStatic)) {
+          context.report({
+            messageId: 'unbound',
+            node
+          });
+        }
+      }
+    };
+  }
+});
+
+function isDangerousMethod(symbol: ts.Symbol, ignoreStatic: boolean) {
+  const { valueDeclaration } = symbol;
+
+  switch (valueDeclaration.kind) {
+    case ts.SyntaxKind.MethodDeclaration:
+    case ts.SyntaxKind.MethodSignature:
+      return !(
+        ignoreStatic &&
+        tsutils.hasModifier(
+          valueDeclaration.modifiers,
+          ts.SyntaxKind.StaticKeyword
+        )
+      );
+  }
+
+  return false;
+}
+
+function isSafeUse(node: TSESTree.BaseNode): boolean {
+  const parent = node.parent!;
+
+  switch (parent.type) {
+    case 'IfStatement':
```

Comment:
you should use `AST_NODE_TYPES.x` instead of raw strings

Reply:
Why? It's defined by the ESTree spec, so it will never change. I don't think longer variable makes sense. It makes readability bad.

Reply:
One benefit of using `AST_NODE_TYPES.x` is that it can help prevent typos.

Reply:
Don't worry about typo. TS language server suggests valid string literals and does type check.

Reply:
It makes refactoring easier because it means there are less raw string definitions in our code base.
It also means that if (for some reason) the type of the `type` prop gets widened to just `string`, you can still catch typos.

Reply:
The biggest thing though is just consistency.
It makes it easy to tell the difference between checking for a token type, and  checking for a node type.

Also if we're doing something like this pattern:
```TS
const allowedNodeTypes = new Set(['IfStatement', ...]);
```

Then it prevents typos and will cause typescript to strictly type the set:
![image](https://user-images.githubusercontent.com/7462525/53038947-c8009000-3432-11e9-918a-06c7b6be59f9.png)

Reply:
> It makes refactoring easier because it means there are less raw string definitions in our code base.

How does it makes refactoring easier?

> It also means that if (for some reason) the type of the `type` prop gets widened to just `string`, you can still catch typos.

How does it widen the type?

> Also if we're doing something like this pattern:

Type parameters are enough: `new Set<NodeType>(['IfStatement', ...]);`
We don't need repeat the noisy prefix for every element.

> It makes it easy to tell the difference between checking for a token type, and checking for a node type.

I have never encountered the situation I want to separate those.
Why `Node` and `Token` types are not enough?

---

`AST_NODE_TYPES` prefix is really noisy to me, but I cannot find no pros.

Reply:
> How does it make refactoring easier?

If we ever rename a node (i.e. an enum member), you can trivially leverage the TS language service to do the refactor. i.e in vscode: click on member, press F2, type in new name, press enter, and *all* usages are renamed - no manual work required.

With strings we have to do a search for `'TypeName'`, and then go through each usage and double check it to make sure that it's a node type comparison, and only then can you replace it.
Also if for some reason there's a clash between a node type and a token type name (i.e. `Identifier`), then you have to take the extra step of interrogating the code to make sure that you're renaming a node usage.

Raw strings are very, very error prone in a refactor due to their complete and utter ambiguity.
Enums are not in the least bit error prone because they are a strictly defined type.

> How does it widen the type?

Typescript can sometimes widen the type definitions if your code gets too complex, or if you do some "javascripty" things which aren't necessarily typesafe.
But there's also cases like I mentioned:

```ts
enum Foo {
  a = 'a',
  b = 'b',
}
const stringSet = new Set(['a', 'b']); // type === Set<string>
const enumSet = new Set([Foo.a, Foo.b]); // type === Set<Foo>
```

There are cases with generics where things can be widened because it's generics.
Using the enum member means you cannot widen the type unless you're doing something like
```
const newName = `${AST_NODE_TYPES.Identifier}` // type === string
```

If you use the enum, then you can be 100% certain that you are never misspelling the type name.

> I have never encountered the situation I want to separate those. Why `Node` and `Token` types are not enough?

When I did the refactor I actually found a few places in the codebase that were ambiguous, and it wasn't until I had 100% nailed down the types that I was able to tell which one it was supposed to be (and even then I could only tell via hovering over the variables to see what their type was).

Which leads me back to my original point that raw strings are very ambiguous, where as enum members are not.

Note also that you cannot accidentally compare a token type to a node type when using enums:
```TS
enum NodeType {
    a = 'a',
    b = 'b',
}
enum TokenType {
    a = 'a',
    c = 'c',
}

function Foo(node: NodeType, token: TokenType) {
    if (node === TokenType.a) { } // errors because enums don't match
    if (token === NodeType.a) { } // errors because enums don't match
}
```
[repl](https://www.typescriptlang.org/play/#src=enum%20NodeType%20%7B%0D%0A%20%20%20%20a%20%3D%20'a'%2C%0D%0A%20%20%20%20b%20%3D%20'b'%2C%0D%0A%7D%0D%0Aenum%20TokenType%20%7B%0D%0A%20%20%20%20a%20%3D%20'a'%2C%0D%0A%20%20%20%20c%20%3D%20'c'%2C%0D%0A%7D%0D%0A%0D%0Afunction%20Foo(node%3A%20NodeType%2C%20token%3A%20TokenType)%20%7B%0D%0A%20%20%20%20if%20(node%20%3D%3D%3D%20TokenType.a)%20%7B%0D%0A%0D%0A%20%20%20%20%7D%0D%0A%20%20%20%20if%20(token%20%3D%3D%3D%20NodeType.a)%20%7B%0D%0A%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A)

> `AST_NODE_TYPES` prefix is really noisy to me, but I cannot find pros.

Sure, it is a little bit of noise, but the pros are:
- It's always 100% typesafe.
- It's "idiot proof" - there's no way to accidentally widen the type and introduce spots that aren't checked for typos.
- it's very, very simple to refactor names if we want.
- it keeps us consistent across the entire monorepo.

Reply:
Thank you for explaining. But you looks to have invalid assumption.

> If we ever rename a node (i.e. an enum member),

The constants of node types defined by ESTree will not change. If it was changed, it will break several tools such as ESLint, Babel, webpack, etc...

> Set example

As I mentioned, we don't need to repeat the prefix.

> Note also that you cannot accidentally compare a token type to a node type when using enums:

`"IfStatement"` is `IfStatement`. The type value describes meaning itself. We really don't need to separate those. The type check reports even if we compare node types and token types without enums.

Reply:
Personally I prefer the `AST_NODE_TYPES.x` behavior to get better IDE features when writing the `case` statements. Specifically auto-completion and go-to-definition are nice here. I'll switch to that for now, and can switch back if there's consensus?